### PR TITLE
Avoid exception when minion does not respond using syndics

### PR DESCRIPTION
**Pull request by sjafferali**
### What does this PR do?
We have several issues reporting that salt is throwing exception when the minion does not respond. This change avoid the exception adding a default data to the minion when it fails to respond. This patch is based on the changes that @jvanz submitted in https://github.com/saltstack/salt/pull/53159 with a few minor changes made in response to the requested comments. 

### What issues does this PR fix or reference?
Issues #46876 #48509 #50238
bsc#1135507
Also PRs https://github.com/saltstack/salt/pull/53159 and https://github.com/saltstack/salt/pull/54725

### Previous Behavior
```
salt -b 50 -L server1.domain.com,fake.domain.com test.ping
```

Fail with
```
[ERROR ] An un-handled exception was caught by salt's global exception handler:
KeyError: u'ret'
Traceback (most recent call last):
File "/usr/bin/salt", line 10, in
salt_main()
File "/usr/lib/python2.7/site-packages/salt/scripts.py", line 485, in salt_main
client.run()
File "/usr/lib/python2.7/site-packages/salt/cli/salt.py", line 65, in run
self._run_batch()
File "/usr/lib/python2.7/site-packages/salt/cli/salt.py", line 283, in _run_batch
for res in batch.run():
File "/usr/lib/python2.7/site-packages/salt/cli/batch.py", line 322, in run
ret[minion] = data['ret']
KeyError: u'ret'
Traceback (most recent call last):
File "/usr/bin/salt", line 10, in
salt_main()
File "/usr/lib/python2.7/site-packages/salt/scripts.py", line 485, in salt_main
client.run()
File "/usr/lib/python2.7/site-packages/salt/cli/salt.py", line 65, in run
self._run_batch()
File "/usr/lib/python2.7/site-packages/salt/cli/salt.py", line 283, in _run_batch
for res in batch.run():
File "/usr/lib/python2.7/site-packages/salt/cli/batch.py", line 322, in run
ret[minion] = data['ret']
KeyError: u'ret'
```
### New Behavior

```
salt --out=raw -b 2 -L germ109.suse.cz,fake.domain.cz test.ping
{u'germ109.suse.cz': True}
{u'fake.domain.cz': u'Minion did not return. [Failed]'}
```

### Tests written?

No. This PR addresses a very specific failure scenario. A command would need to be run on a master of masters, with a syndic connected to it, and a subsequent minion connected to that syndic, which I have not been able to trigger using the current testing framework. tests/integration/cli/test_batch.py runs on a master rather than a master of masters, which prevents running into this error due to additional safety checks that are hit first (you must target a non-existent minion to reach the failure condition). See linked issues for further details. 

### Commits signed with GPG?

No

### CODE REVIEW

**Comment by sjafferali** [comment]
@xeacott @DmitryKuzmenko - Could you possibly assist with reviewing this PR for the issues linked above?

**Comment by Oloremo** [comment]
mind to rebasing it on top of the current master?

**Comment by DmitryKuzmenko** [comment]
By the code this PR looks reasonable for me. Anywhere I really like to see a test for this. @sjafferali is there any chance for you to write that? I'll be happy to help you with that here or in Slack?

**Comment by waynew** [comment]
@DmitryKuzmenko we've got some tests now! 🚀 

Also ended out discovering another potential bug and fixing that while I was at it 😂

**Comment by waynew** [comment]
re-run all

**Review by garethgreenaway — APPROVED** [review]

**Review by twangboy — APPROVED** [review]

File: salt/cli/batch.py [hunk 1]
```diff
@@ -116,10 +116,9 @@ def run(self):
         bnum = self.get_bnum()
         # No targets to run
         if not self.minions:
-            return
+            raise StopIteration("No minions matched target")
```
Comment (s0undt3ch) [inline salt/cli/batch.py:119]: Functions which `yield`, after Python 3.7 throw RuntimeErrors
https://docs.python.org/3.7/whatsnew/3.5.html?highlight=stopiteration#pep-479-change-stopiteration-handling-inside-generators

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

**Comment by sagetherage** [comment]
I would still like to see this get into the code base, but with changes requested and not passing tests I am moving it to the AL point release

**Comment by sagetherage** [comment]
This will not be ready to be merged in for the release Aluminium point release and scoping until changes have been addressed

**Comment by cyrinux** [comment]
Hi, 
the issue is still present, and this PR would be useful if merge.

Can we make this PR rebase on the upstream and merge please? 🙏🏻 

Is there any blocker actually?

I was ready to send a PR for it but I finally fall on this one.

Cheers

**Comment by Ch3LL** [comment]
Looks like we need the following:

**Comment by dwoz** [comment]
Is this still in progress?

**Comment by dwoz** [comment]
Closing this due to inactivity. Anyone should feel free to re-open it if they want to see it through to the end in one release cycle.

@sjafferali This needs to be rebased and @s0undt3ch comment should be addressed. If you have the bandwidth to try and get it into a release please re-open.

**Pull request closed without merging**